source/codegen/metadata/nidcpower/functions.py

```diff
@@ -0,0 +1,2985 @@
+        'type': 'ViSession'
+      },
+      {
+        'name': 'temperature',
+        'direction': 'out',
+        'type': 'ViReal64'
+      }
+    ],
+    'returns': 'ViStatus'
+  },
+  'ImportAttributeConfigurationFile': {
+    'parameters': [
+      {
+        'name': 'vi',
+        'direction': 'in',
+        'type': 'ViSession'
+      },
+      {
+        'name': 'filePath',
+        'direction': 'in',
+        'type': 'ViConstString'
+      }
+    ],
+    'returns': 'ViStatus'
+  },
+  'InitExtCal': {
+    'parameters': [
+      {
+        'name': 'resourceName',
+        'direction': 'in',
+        'type': 'ViRsrc'
+      },
+      {
+        'name': 'password',
+        'direction': 'in',
+        'type': 'ViConstString'
+      },
+      {
+        'name': 'vi',
+        'direction': 'out',
+        'type': 'ViSession'
+      }
+    ],
+    'returns': 'ViStatus'
+  },
+  'InitializeWithChannels': {
+    'init_method' : True,
+    'parameters': [
+      {
+        'name': 'resourceName',
+        'direction': 'in',
+        'type': 'ViRsrc'
+      },
+      {
+        'name': 'channels',
+        'direction': 'in',
+        'type': 'ViConstString'
+      },
+      {
+        'name': 'reset',
+        'direction': 'in',
+        'type': 'ViBoolean'
+      },
+      {
+        'name': 'optionString',
+        'direction': 'in',
+        'type': 'ViConstString'
+      },
+      {
+        'name': 'vi',
+        'direction': 'out',
+        'type': 'ViSession'
+      }
+    ],
+    'returns': 'ViStatus'
+  },
+  'Initiate': {
+    'parameters': [
+      {
+        'name': 'vi',
+        'direction': 'in',
+        'type': 'ViSession'
+      }
+    ],
+    'returns': 'ViStatus'
+  },
+  'LockSession': {
+    'parameters': [
+      {
+        'name': 'vi',
+        'direction': 'in',
+        'type': 'ViSession'
+      },
+      {
+        'name': 'callerHasLock',
+        'direction': 'out',
+        'type': 'ViBoolean'
+      }
+    ],
+    'returns': 'ViStatus'
+  },
+  'Measure': {
+    'parameters': [
+      {
+        'name': 'vi',
+        'direction': 'in',
+        'type': 'ViSession'
+      },
+      {
+        'name': 'channelName',
+        'direction': 'in',
+        'type': 'ViConstString'
+      },
+      {
+        'name': 'measurementType',
+        'direction': 'in',
+        'type': 'ViInt32',
+        'enum': 'MeasurementTypes'
+      },
+      {
+        'name': 'measurement',
+        'direction': 'out',
+        'type': 'ViReal64'
+      }
+    ],
+    'returns': 'ViStatus'
+  },
+  'MeasureMultiple': {
+    'parameters': [
+      {
+        'name': 'vi',
+        'direction': 'in',
+        'type': 'ViSession'
+      },
+      {
+        'name': 'channelName',
+        'direction': 'in',
+        'type': 'ViConstString'
+      },
+      {
+        'name': 'voltageMeasurements',
+        'direction': 'out',
+        'type': 'ViReal64[]',
+        'size': {
+          'mechanism': 'source-code'
```

Comment:
Usually when the mechanism is `source-code`, there's a `value` entry with the source code to run. Try to include what the source code would look like in C++. See the niscope functions as an example.

Reply:
@jonlugoatx, This API needs to be custom implemented, so we have planned to make changes in metadata and do the custom implementation as part of subsequent PR's. I would probably comment-out this function in the metadata for this PR.

Reply:
Don't comment out the metadata. Instead, do this to be consistent with [NI-SCOPE metadata](https://github.com/ni/grpc-device/blob/main/source/codegen/metadata/niscope/functions.py#L1193):
```python
'size': {
  'mechanism': 'custom-code',
  'value': 'number_of_channels'
}
```
Explanation:
The custom code you need to implement is a call to `niDCPower_ParseChannelCount` (see the [nimi-python metadata](https://github.com/ni/nimi-python/blob/master/src/nidcpower/metadata/functions.py#L1983) & [helper implementation](https://github.com/ni/nimi-python/blob/master/generated/nidcpower/nidcpower/session.py#L3531)). Then your custom code would pass the output parameter of that call as the `voltageMeasurements` argument when calling `niDCPower_MeasureMultiple`. The output variable of that function call should go in the `value` field of the metadata.

This is very similar to the NI-SCOPE plugin where scope needs to call `niScope_ActualNumWfms`. We considered adding a gRPC codegen feature to specify custom functions to call but cut that codegen feature. If we implement that feature, then we would need the value in the size metadata to be the C++ code to write for the argument in the function call.

When you make this metadata change, you don't need to provide a custom implementation of the function until you add `'codegen_method': 'CustomCode'` before the `'parameters'` array.